___
# <span style="color:darkgreen">📊 Retail Analytics: Walmart Case Study</span>
___

### Project Summary

**Domain:** Finance

**Objective:** This project seeks to analyze Walmart’s transactional data from the past five years to generate actionable insights into customer behavior, sales trends, product performance, and operational efficiency.


**Business Problem :** The problem is to analyze Walmart’s historical transactional dataset covering customer, product, and logistics information; in order to generate insights that improve profitability, enhance customer satisfaction, and streamline operational efficiency.

**Data set:** The dataset contains five years of Walmart transactions, including customer demographics, order details, product attributes, and sales logistics.  
***Data source :*** https://www.kaggle.com/datasets/saadabdurrazzaq/walmart-retail-data?resource=download

### Step 1. Data Loading and Initial Overview

In [2]:
#Importing pandas libraries
import pandas as pd
import numpy as np
#Display setting for pandas
pd.set_option('display.max_columns', None)              #show all columns
pd.set_option('display.float_format','{:.2f}'.format)    #format all decimals to 2 places

print("Libraries imported successfully!")

Libraries imported successfully!


**1.1 Dataset Overview**   
___
This dataset captures five years of Walmart retail transactions, providing details on customers, orders, products, and logistics. Key information includes:
- Customer data: demographics (age, name, segment), location (city, state, ZIP).
- Order details: order ID, date, priority, quantity, discounts, and shipping (mode, cost, dates).
- Product attributes: category, sub-category, name, container type, unit price, and base margin.
- Performance metrics: sales, profit, and regional breakdowns.
  
Together, these fields enable analysis of customer behavior, product performance, profitability, and operational efficiency across Walmart’s retail ecosystem.


In [3]:
#Load the dataset
df = pd.read_csv('../data/Walmart Retail Data.csv', encoding='latin-1')
# Display dataset dimensions
print("Dataset loaded successfully!")
print("="*70)
print(f"Dataset Shape: {df.shape[0]:,} rows × {df.shape[1]} columns")

Dataset loaded successfully!
Dataset Shape: 8,399 rows × 25 columns


**1.2 First 5 Rows of Dataset**   
To check the initial view of data inluding columns, data types and structure

In [6]:
# Display the first 5 rows of the dataset
print("First 5 rows of the dataset:")
display(df.head())


First 5 rows of the dataset:


,City,Customer Age,Customer Name,Customer Segment,Discount,Number of Records,Order Date,Order ID,Order Priority,Order Quantity,Product Base Margin,Product Category,Product Container,Product Name,Product Sub-Category,Profit,Region,Row ID,Sales,Ship Date,Ship Mode,Shipping Cost,State,Unit Price,Zip Code
0,McKeesport,NaN,Jessica Myrick,Small Business,0.10,1,01-01-2012,28774,High,32,0.68,Office Supplies,Small Box,"Perma STOR-ALL™ Hanging File Box, 13 1/8""W x 1...",Storage & Organization,-111.80,East,4031,180.36,02-01-2012,Regular Air,4.69,Pennsylvania,5.98,15131
1,Bowie,NaN,Matt Collister,Home Office,0.08,1,01-01-2012,13729,Not Specified,9,NaN,Office Supplies,Large Box,Safco Industrial Wire Shelving,Storage & Organization,-342.91,East,1914,872.48,03-01-2012,Express Air,35.00,Maryland,95.99,20715
2,Napa,NaN,Alan Schoenberger,Corporate,0.00,1,02-01-2012,37537,Low,4,0.56,Furniture,Jumbo Drum,Hon 4070 Series Pagoda™ Armless Upholstered St...,Chairs & Chairmats,-193.08,West,5272,1239.06,02-01-2012,Delivery Truck,48.80,California,291.73,94559
3,Montebello,NaN,Elizabeth Moffitt,Consumer,0.08,1,02-01-2012,44069,Critical,43,0.39,Office Supplies,Wrap Bag,White GlueTop Scratch Pads,Paper,247.79,West,6225,614.80,02-01-2012,Regular Air,1.97,California,15.04,90640
4,Napa,NaN,Alan Schoenberger,Corporate,0.07,1,02-01-2012,37537,Low,43,0.69,Furniture,Jumbo Drum,Hon Valutask™ Swivel Chairs,Chairs & Chairmats,-1049.85,West,5273,4083.19,04-01-2012,Delivery Truck,45.00,California,100.98,94559


**1.3 Last 5 Rows of Dataset**   
To ensure complete data have been loaded

In [5]:
# Display the last 5 rows of the dataset
print("Last 5 rows of the dataset:")
df.tail()

Last 5 rows of the dataset:


,City,Customer Age,Customer Name,Customer Segment,Discount,Number of Records,Order Date,Order ID,Order Priority,Order Quantity,Product Base Margin,Product Category,Product Container,Product Name,Product Sub-Category,Profit,Region,Row ID,Sales,Ship Date,Ship Mode,Shipping Cost,State,Unit Price,Zip Code
8394,Fairfield,95.00,Tony Molinari,Corporate,0.10,1,30-12-2015,50950,Not Specified,35,0.59,Office Supplies,Small Box,Tenex Personal Project File with Scoop Front D...,Storage & Organization,-15.07,West,7142,448.10,30-12-2015,Express Air,4.51,California,13.48,94533
8395,Harker Heights,95.00,Matt Hagelstein,Home Office,0.09,1,30-12-2015,25542,Low,37,0.39,Office Supplies,Wrap Bag,"Black Print Carbonless 8 1/2"" x 8 1/4"" Rapid M...",Paper,-18.66,Central,3583,257.46,30-12-2015,Express Air,4.23,Texas,7.28,76543
8396,Riverview,95.00,Theresa Swint,Consumer,0.10,1,30-12-2015,45127,Medium,10,0.37,Office Supplies,Wrap Bag,Binder Clips by OIC,Rubber Bands,-1.29,South,6361,14.15,30-12-2015,Regular Air,0.70,Florida,1.48,33569
8397,Nicholasville,95.00,Maribeth Yedwab,Home Office,0.09,1,30-12-2015,49344,Low,1,0.83,Office Supplies,Medium Box,Martin Yale Chadless Opener Electric Letter Op...,"Scissors, Rulers and Trimmers",-745.20,South,6916,803.33,30-12-2015,Regular Air,24.49,Kentucky,832.81,40356
8398,Nicholasville,95.00,Maribeth Yedwab,Home Office,0.00,1,30-12-2015,49344,Low,31,0.68,Technology,Small Box,Belkin 105-Key Black Keyboard,Computer Peripherals,27.85,South,6915,672.93,30-12-2015,Regular Air,4.00,Kentucky,19.98,40356


**1.4 Dataset Information**  
To retrieve information about the data such as data types, number of rows and colums

In [15]:
# Display dataset information
print("Dataset Information:")
print("="*70)
print("\nDataset Info:")
df.info()


Dataset Information:

Dataset Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8399 entries, 0 to 8398
Data columns (total 25 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   City                  8399 non-null   object 
 1   Customer Age          7496 non-null   float64
 2   Customer Name         8399 non-null   object 
 3   Customer Segment      8399 non-null   object 
 4   Discount              8399 non-null   float64
 5   Number of Records     8399 non-null   int64  
 6   Order Date            8399 non-null   object 
 7   Order ID              8399 non-null   int64  
 8   Order Priority        8399 non-null   object 
 9   Order Quantity        8399 non-null   int64  
 10  Product Base Margin   8336 non-null   float64
 11  Product Category      8399 non-null   object 
 12  Product Container     8399 non-null   object 
 13  Product Name          8399 non-null   object 
 14  Product Sub-Category  8399 non-null 

**1.5 Statistical Summary**   
The Statistical Summary provides average, minimum, maximum, std deviations, quartiles to better understand data with high level overview

In [17]:
# Statistical summary of numerical columns
print("\nSummary Statistics:")
display(df.describe(include='number'))



Summary Statistics:


,Customer Age,Discount,Number of Records,Order ID,Order Quantity,Product Base Margin,Profit,Row ID,Sales,Shipping Cost,Unit Price,Zip Code
count,7496.00,8399.00,8399.00,8399.00,8399.00,8336.00,8399.00,8399.00,8399.00,8399.00,8399.00,8399.00
mean,54.54,0.05,1.00,29965.18,25.57,0.51,181.18,4200.00,1775.88,12.84,89.35,52839.14
std,9.52,0.03,0.00,17260.88,14.48,0.14,1196.65,2424.73,3585.05,17.26,290.35,28509.54
min,41.00,0.00,1.00,3.00,1.00,0.35,-14140.70,1.00,2.24,0.49,0.99,1001.00
25%,47.00,0.02,1.00,15011.50,13.00,0.38,-83.31,2100.50,143.19,3.30,6.48,30337.00
50%,53.00,0.05,1.00,29857.00,26.00,0.52,-1.50,4200.00,449.42,6.07,20.99,52732.00
75%,61.00,0.08,1.00,44596.00,38.00,0.59,162.75,6299.50,1709.32,13.99,85.99,77577.00
max,95.00,0.25,1.00,59973.00,50.00,0.85,27220.69,8399.00,89061.05,164.73,6783.02,99362.00


**1.6 Data Quality Checks**   
To check how many missing values and duplicate items exist in data

In [18]:
# Check for missing values in all columns
print("Missing Values Summary:")
missing_values = df.isnull().sum()
print(missing_values)
print("="*70)

# Columns with missing values
if missing_values.sum() > 0:
    print("\nColumns with missing values:")
    print(missing_values[missing_values > 0])
else:
    print("\nNo missing values found in the dataset.")

Missing Values Summary:
City                      0
Customer Age            903
Customer Name             0
Customer Segment          0
Discount                  0
Number of Records         0
Order Date                0
Order ID                  0
Order Priority            0
Order Quantity            0
Product Base Margin      63
Product Category          0
Product Container         0
Product Name              0
Product Sub-Category      0
Profit                    0
Region                    0
Row ID                    0
Sales                     0
Ship Date                 0
Ship Mode                 0
Shipping Cost             0
State                     0
Unit Price                0
Zip Code                  0
dtype: int64

Columns with missing values:
Customer Age           903
Product Base Margin     63
dtype: int64


In [19]:
# Check for duplicate rows
duplicates = df.duplicated().sum()
print(f"Number of duplicate rows: {duplicates}")

if duplicates > 0:
    print(f"\nWarning: {duplicates} duplicate row(s) detected!")
else:
    print("\nNo duplicate rows found in the dataset.")

Number of duplicate rows: 0

No duplicate rows found in the dataset.


**1.7 Key Insights**  
ddasd

<div style="border:1px solid #ccc; padding:10px; border-radius:5px; background-color:lightyellow">
<h4>🔑 Key Insights provided below</h4>
<ul>
  <li>Data consist of 8,399 rows and 25 columns.</li>
  <li>Average age of customer is 55 (Range 41-95).</li>
  <li>Average shipping cost is 13 (Range 0.50-165).</li>
  <li>Average Product Margine is 0.51 (Range 0.35-0.85).</li>
  <li>Customer Age has 903 missing values (10.75%).</li>
  <li>Product Base Margin has 63 missing values(0.75%).</li>
  <li>No duplicate rows find.</li>
  <li>Sales and Profit columns are complete.</li>
</ul>
</div>
